using_restricted_array.rb

```diff
@@ -7,41 +7,117 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  element = 0
+  while element.class == Integer
+    element = array[index]
+    index += 1
+  end
+  index - 1
 end
 
+
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  element = 0
+  while element.class == Integer
+    element = array[index]
+    index += 1
+    print "#{element} "
+  end
+  puts
+  index - 1
 end
 
+
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  element = nil
+  index = 0
+  while element != value_to_find
+    element = array[index]
+    index += 1
+    return false if index > length
+  end
+  return index - 1
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
+
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  largest = array[index]
+  while index < length
+    if array[index] > largest
+      largest = array[index]
+    end
+    index += 1
+  end
+  largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  largest = array[index]
+  while index > length
+    if array[index] > largest
```

Comment:
Line 68 should be `if array[length] < largest # really a better variable name would be smallest`.